sdk/core/Azure.Core/src/PipelineContext.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Azure.Core
+{
+    /// <summary>
+    /// Metadata describing the flow of the HttpMessage through the pipeline
+    /// in a given invocation.
+    /// </summary>
+    public class PipelineContext
+    {
+        internal PipelineContext(DateTimeOffset createdOn)
+        {
+            CreatedOn = createdOn;
+        }
+
+        /// <summary>
+        /// The time this HttpMessage was created.
+        /// </summary>
+        public DateTimeOffset CreatedOn { get; }
+
+        private int? _retryAttempt;
+
+        /// <summary>
+        /// The number indicating which retry attempt is currently in progress.
+        /// </summary>
+        public int RetryAttempt
+        {
+            get
+            {
+                if (_retryAttempt == null)
+                {
+                    throw new InvalidOperationException("Tried to read RetryAttempt when it did not have a value. " +
```

Comment:
- This ensures that someone won't try to depend on this value downstream from a custom retry policy that forgets to set the attempt number.
- The assumption is that any issues would be caught at development time.

Reply:
I don't think we can rely on this. We'd need to extract the setting of the context into a policy that goes after RetryPolicy.

Reply:
Why is the extra policy required?

Reply:
The reason for having the extra policy is that we would be able to ensure the context values are set correctly. Having a runtime check doesn't guarantee that the context is set correctly.

Reply:
Makes sense.  It might be worth exploring if there's any other way to approach that that avoids adding an extra policy.  Or maybe it is better to not rely on other policies to set Retry/Processing/PipelineContext, in which case it might make sense to consider what else could go on the context and where it would be best to put the other policy.  For example, we currently do [runtime validation that the Response is set](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/HttpMessage.cs#L54) -- would it be better to do that in the extra policy?  Would we always want to have that policy in the pipeline, or only when there are policies that rely on RetryContext being set?

Reply:
I think we'd always want this extra policy to ensure the context is set regardless of whether the user replaces the retry policy or not. If it turns out this is not feasible, then I think we could use some sort of runtime validation like you have suggested.

Reply:
I had understood from Krzysztof at some point that it was a goal to try not to add new required policies to the pipeline, but you may want to double check that with him and how concerned about that he still is.

Reply:
Pinged @KrzysztofCwalina

Reply:
> This ensures that someone won't try to depend on this value downstream from a custom retry policy that forgets to set the attempt number.

I don't think we should be throwing from this property. It should start with  the first value (1 I think) and we should assume that the retry policy increments it properly.

Reply:
Discussed offline - @KrzysztofCwalina suggested just leaving the context fully mutable and not adding any kind of validation. It it is the user's responsibility to update the context if they override the policy.

Reply:
Thanks for following up on that Josh!